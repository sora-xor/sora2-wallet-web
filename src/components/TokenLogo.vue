<template>
  <div class="logo">
    <span :class="iconClasses" :style="iconStyles" />
    <nft-token-logo v-if="isNft" class="asset-logo__nft-image" :class="iconClasses" :asset="token" />
  </div>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { api } from '../api';
import { LogoSize, ObjectInit } from '../consts';
import { getter } from '../store/decorators';

import NftTokenLogo from './NftTokenLogo.vue';
import TranslationMixin from './mixins/TranslationMixin';

import type { WhitelistIdsBySymbol } from '../types/common';
import type { Asset, AccountAsset, Whitelist, WhitelistItem } from '@sora-substrate/util/build/assets/types';

@Component({
  components: {
    NftTokenLogo,
  },
})
export default class TokenLogo extends Mixins(TranslationMixin) {
  @getter.account.whitelist private whitelist!: Whitelist;
  @getter.account.whitelistIdsBySymbol private whitelistIdsBySymbol!: WhitelistIdsBySymbol;

  @Prop({ type: String, default: '' }) readonly tokenSymbol!: string;
  @Prop({ type: Object, default: ObjectInit }) readonly token!: Nullable<AccountAsset | Asset>;
  @Prop({ type: String, default: LogoSize.MEDIUM, required: false }) readonly size!: LogoSize;
  @Prop({ default: false, type: Boolean }) readonly withClickableLogo!: boolean;

  get isNft(): boolean {
    return !!this.token && api.assets.isNft(this.token);
  }

  get assetAddress(): Nullable<string> {
    return this.tokenSymbol ? this.whitelistIdsBySymbol[this.tokenSymbol] : (this.token || {}).address;
  }

  get whitelistedItem(): Nullable<WhitelistItem> {
    if (!(this.token || this.tokenSymbol)) {
      return null;
    }
    const address = this.assetAddress;
    if (!address) {
      return null;
    }
    return this.whitelist[address];
  }

  get iconStyles(): object {
    const asset = this.whitelistedItem;
    if (!asset) {
      return {};
    }
    return {
      'background-size': '100%',
      'background-image': `url("${asset.icon}")`,
    };
  }

  get iconClasses(): Array<string> {
    const questionMark = 's-icon-notifications-info-24';
    const tokenLogoClass = 'asset-logo';
    const classes = [tokenLogoClass];

    if (!this.assetAddress) {
      classes.push(questionMark);
    } else if (!this.whitelistedItem) {
      classes.push(this.isNft ? 'asset-logo-nft' : questionMark);
    }

    classes.push(`${tokenLogoClass}--${this.size.toLowerCase()}`);

    if (this.withClickableLogo) {
      classes.push('asset-logo--clickable');
    }

    return classes;
  }
}
</script>

<style lang="scss" scoped>
$token-background-color: var(--s-color-base-on-accent);
$token-color: var(--s-color-base-content-tertiary);

.logo {
  position: relative;
}
.asset-logo {
  &__nft-image {
    border-radius: 50%;
    object-fit: cover;
    position: absolute !important;
    top: 0;
    left: 0;
  }

  &--clickable {
    cursor: pointer;
  }
}

@mixin token-logo-size($size: '') {
  $className: 'asset-logo';
  @if ($size == 'mini') {
    $size-px: 16px;
    $classNameMini: '#{$className}--mini';
    @include element-size($classNameMini, $size-px);
    .#{$classNameMini} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 12px;
      color: var(--s-color-base-content-tertiary);
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 6px;
      }
    }
  } @else if ($size == 'small') {
    $size-px: 24px;
    $classNameSmall: '#{$className}--small';
    @include element-size($classNameSmall, $size-px);
    .#{$classNameSmall} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 18px;
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 8px;
        font-weight: 800;
      }
    }
  } @else if ($size == 'medium') {
    $size-px: 32px;
    $classNameMedium: '#{$className}--medium';
    @include element-size($classNameMedium, $size-px);
    .#{$classNameMedium} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 24px;
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 12px;
        font-weight: 800;
      }
    }
  } @else if ($size == 'big') {
    $size-px: var(--s-size-medium);
    $classNameBig: '#{$className}--big';
    @include element-size($classNameBig, $size-px);
    .#{$classNameBig} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 32px;
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 14px;
        font-weight: 800;
      }
    }
  } @else if ($size == 'bigger') {
    $size-px: 48px;
    $classNameBig: '#{$className}--bigger';
    @include element-size($classNameBig, $size-px);
    .#{$classNameBig} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 32px;
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 14px;
        font-weight: 800;
      }
    }
  } @else if ($size == 'large') {
    $size-px: 80px;
    $classNameLarge: '#{$className}--large';
    @include element-size($classNameLarge, $size-px);
    .#{$classNameLarge} {
      @include asset-logo-styles;
      line-height: $size-px;
      font-size: 64px;
      &.#{$className}-nft::before {
        content: 'NFT';
        font-size: 28px;
        font-weight: 800;
      }
    }
  }
}

@include token-logo-size('mini');
@include token-logo-size('small');
@include token-logo-size('medium');
@include token-logo-size('big');
@include token-logo-size('bigger');
@include token-logo-size('large');
</style>
