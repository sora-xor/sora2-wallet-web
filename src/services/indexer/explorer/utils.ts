import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
