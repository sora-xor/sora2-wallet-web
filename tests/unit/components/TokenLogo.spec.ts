import { useDescribe, useShallowMount, useVuex } from '../../utils';
import TokenLogo from '@/components/TokenLogo.vue';
import { MOCK_ACCOUNT_ASSETS, MOCK_ACCOUNT_ASSETS_NFT } from '../../utils/mock';

const createStore = (mutations) =>
  useVuex({
    router: {
      mutations,
    },
  });

useDescribe('TokenLogo.vue', TokenLogo, () => {
  it('should render nft-token-logo when asset.content exists', () => {
    const wrapper = useShallowMount(TokenLogo, {
      propsData: {
        token: MOCK_ACCOUNT_ASSETS_NFT[0],
      },
    });
    const nftComponent = wrapper.find('.asset-logo-nft');

    expect(nftComponent.exists()).toBeTrue();
  });

  it('should call navigate() mutation when prop is passed and logo is clicked', () => {
    const mutations = {
      navigate: jest.fn(),
    };

    const wrapper = useShallowMount(TokenLogo, {
      store: createStore(mutations),
      propsData: {
        token: MOCK_ACCOUNT_ASSETS[0],
        withClickableLogo: true,
      },
    });

    const element = wrapper.find('.asset-logo');

    element.trigger('click');

    expect(mutations.navigate).toHaveBeenCalled();
  });
});
