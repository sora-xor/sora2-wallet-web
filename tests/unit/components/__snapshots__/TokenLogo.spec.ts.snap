// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenLogo.vue should render correctly 1`] = `
<fragment-stub
  name="12cb0e4f4ff"
>
  <span
    class="s-icon-notifications-info-24 asset-logo asset-logo--medium"
  />
   
  <!---->
</fragment-stub>
`;

exports[`TokenLogo.vue should render nft-token-logo when asset.content exists 1`] = `
<fragment-stub
  name="7e07c14e05"
>
  <span
    class="asset-logo-nft asset-logo asset-logo--medium"
  />
   
  <nft-token-logo-stub
    asset="[object Object]"
    class="asset-logo asset-logo__nft-image asset-logo-nft asset-logo asset-logo--medium"
  />
</fragment-stub>
`;
